File: backend/ExamScheduleManagement/src/main/java/org/examschedulemanagement/Entities/Exam.java
package org.examschedulemanagement.Entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Exam {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id ;

    public Exam(Semester semestere, Session session, Course course, TypeExam typeExam, LocalDate date, LocalDateTime starting_hour, int plannedDuration) {
        this.semestere = semestere;
        this.session = session;
        this.course = course;
        this.typeExam = typeExam;
        this.date = date;
        this.startingHour = starting_hour;
        this.plannedDuration = plannedDuration;
    }
    private Semester semestere;

    private Session session;

    @ManyToOne
    @JoinColumn(name = "course_id")
    private Course course;

    private TypeExam typeExam;

    private LocalDate date;
    private LocalDateTime startingHour;
    private int plannedDuration ;
    private int actualDuration;
    @JsonIgnore
    @OneToMany(mappedBy = "exam",cascade = CascadeType.ALL)
    private List<Surveillance> exam_surveill;


    private String epreuve ;
    private String Pv;
    private String Rapport;
}

File: frontend/exam-schedule/src/types/exam.ts
import { Course } from "@/types/Course";

export interface Exam {
    id?: number;
    semestere: Semester;
    session: Session;
    course: Course;
    typeExam: TypeExam;
    date: string;
    startingHour: string;
    plannedDuration: number;
    actualDuration: number;
    exam_surveill: Surveillance;
    assignments: SalleAssignment[];
    epreuve: string;
    Pv: string;
    Rapport: string;
}

export interface Semester {
    id?: number;
}

export interface Session {
    id?: number;
}

export interface TypeExam {
    id?: number;
    // Add properties as needed
}

export interface Surveillance {
    id?: number;
}

export interface SalleAssignment {
    id?: number;
}

export interface ExamForm1Dto {
    semestere: string;
    session: string;
    course: string;
    typeExam: string;
    date: string;
    starting_hour: string;
}

export interface ExamForm2Dto {
    responseDto: ExamResponseDto;
    plannedDuration: number;
    nbrOfSurv: number;
    random: boolean;
}

export interface ExamResponseDto {
    examForm1Dto: ExamForm1Dto;
    plannedDuration: number;
    nbrStudents: number;
}

export interface ExamForm3Dto {
    responseDto: ExamResponseDto;
    actualDuration: number;
    exam_surveill: Surveillance;
    assignments: SalleAssignment[];
    epreuve: string;
    Pv: string;
    Rapport: string;
}
